src/writers/writer.hh

```diff
@@ -0,0 +1,117 @@
+#ifndef H_ACQUIRE_ZARR_WRITER_V0
+#define H_ACQUIRE_ZARR_WRITER_V0
+
+#ifndef __cplusplus
+#error "This header requires C++20"
+#endif
+
+#include "platform.h"
+#include "device/props/components.h"
+
+#include "../common.hh"
+#include "blosc.compressor.hh"
+
+#include <condition_variable>
+#include <filesystem>
+#include <mutex>
+#include <optional>
+#include <queue>
+#include <thread>
+
+namespace fs = std::filesystem;
+
+namespace acquire::sink::zarr {
+struct Zarr;
+
+struct FileCreator
+{
+    FileCreator() = delete;
+    explicit FileCreator(Zarr* zarr);
+    ~FileCreator() noexcept = default;
+
+    void set_base_dir(const fs::path& base_dir) noexcept;
+    [[nodiscard]] bool create(int n_c,
+                              int n_y,
+                              int n_x,
+                              std::vector<file>& files) noexcept;
+
+  private:
+    fs::path base_dir_;
+    Zarr* zarr_;
+
+    bool create_channel_dirs_(int n_c) noexcept;
+};
+
+struct Writer
+{
+  public:
+    Writer() = delete;
+    Writer(const ImageDims& frame_dims,
+           const ImageDims& tile_dims,
+           uint32_t frames_per_chunk,
+           const std::string& data_root,
+           Zarr* zarr);
+
+    /// Constructor with Blosc compression params
+    Writer(const ImageDims& frame_dims,
+           const ImageDims& tile_dims,
+           uint32_t frames_per_chunk,
+           const std::string& data_root,
+           Zarr* zarr,
+           const BloscCompressionParams& compression_params);
+    virtual ~Writer();
+
+    [[nodiscard]] virtual bool write(const VideoFrame* frame) noexcept = 0;
+    void finalize() noexcept;
+
+    uint32_t frames_written() const noexcept;
+
+  protected:
+    /// Tiling/chunking
+    ImageDims frame_dims_;
+    ImageDims tile_dims_;
+    uint16_t tiles_per_frame_x_;
+    uint16_t tiles_per_frame_y_;
+    SampleType pixel_type_;
+    uint32_t frames_per_chunk_;
+
+    /// Compression
+    std::optional<BloscCompressionParams> blosc_compression_params_;
+    // std::optional<ZstdCompressionParams> zstd_compression_params_; // TODO
+
+    /// Filesystem
+    FileCreator file_creator_;
+    fs::path data_root_;
+    std::vector<file> files_;
+
+    /// Multithreading
+    std::vector<std::vector<uint8_t>> chunk_buffers_;
+    bool* buffers_ready_;
+    std::mutex mutex_;
```

Comment:
What is the mutex being used to guard? I think it's mostly for `chunk_buffers_`, but maybe `buffers_ready_` too?

Could that be communicated in its name? E.g. `buffers_mutex_`.

I often find being specific and explicit about what is intended to be guarded pretty useful, even if it's just in the form of a comment. I really like static analysis for things like this and things like Java's [GuardedBy](https://errorprone.info/bugpattern/GuardedBy), but I'm unaware of something we can use for C++ and that's definitely overkill for this PR.